client.go

```diff
@@ -87,13 +77,37 @@ func (c *Client) NewSession(ctx context.Context, opts *SessionOptions) (*Session
 		HandleMax:      s.handleMax,
 	}
 	log.Debug(1, "TX (NewSession): %s", begin)
-	_ = s.txFrame(begin, nil)
+
+	// we use send to have positive confirmation on transmission
+	send := make(chan encoding.DeliveryState)
+	_ = s.txFrame(begin, send)
 
 	// wait for response
 	var fr frames.Frame
 	select {
 	case <-ctx.Done():
-		// TODO: this will leak s
+		select {
+		case <-send:
+			// begin was written to the network.  assume it was
+			// received and that the ctx was too short to wait for
+			// the ack. in this case we must send an end before we
+			// can delete the session
+			go func() {
+				_ = s.txFrame(&frames.PerformEnd{}, nil)
+				select {
+				case <-c.conn.Done:
+					// conn has terminated
+				case <-time.After(5 * time.Second):
```

Comment:
I think it's worth logging here.

Reply:
I can, but this got me thinking; if we don't receive the ack that the session has ended, we don't know if the peer thinks it's open.  Perhaps in this case we shouldn't delete `s` and just leave it dangling.  This way we won't attempt to reuse the channel.  Thoughts?

Reply:
I thought about this too! 

Maybe that means it's a good idea - I originally got a bit nervous just having state that _never_ expires. Eventually the service, even if it _did_ leave it open, would close the link since it would be idle or completely non-responsive.

Reply:
I've added the logging and moved session deletion to happen on ack only.